doc/contributor/interop.md

```diff
@@ -340,44 +343,54 @@ Not supported.
 
 `value.nil?`
 
+### `HAS_KEYS`
+
+`object.respond_to?(:keys)`
+
 ### `KEYS`
 
-Not supported.
+`value.keys`
 
 ### `READ`
 
-`object[name]`, where name is a `String` or `Symbol` in most cases, or an
+`object[name/index]`, where name is a `String` or `Symbol` in most cases, or an
 integer, or anything else
 
 ### `WRITE`
 
-`object[name] = value`, where name is a `String` or `Symbol` in most cases, or
-an integer, or anything else
+`object[name/index] = value`, where name is a `String` or `Symbol` in most
+cases, or an integer, or anything else
 
 ### `REMOVE`
 
 Not supported.
 
 ## What messages are sent for Ruby syntax on foreign objects
 
-`object[name]` (`#[](name)`) sends `READ`
+TruffleRuby automatically provides these special methods on a foreign object.
+They have priority over methods that the foreign object actually provides.
+
+`object[name/index]` (`#[](name/index)`) sends `READ`.
 
-`object[name] = value` (`#[]=(name, value)`) sends `WRITE`
+`object[name/index] = value` (`#[]=(name/index, value)`) sends `WRITE`.
 
-`object.call(*args)` sends `EXECUTE`
+`object.delete(name/index)` sends `REMOVE`.
 
-`object.nil?` sends `IS_NIL`
+`object.call(*args)` sends `EXECUTE`.
 
-`object.name` sends `INVOKE`
+`object.nil?` sends `IS_NIL`.
 
-`object.name(*args)` sends `INVOKE`
+`object.size` sends `SIZE`.
 
-`object.new(*args)` sends `NEW`
+`object.keys` sends `KEYS` (you may get a foreign array back).
 
-`object.respond_to?` calls `Truffle::Interop.respond_to?(object, message)`,
-which supports `to_a`, `to_ary`, `new`, returning `false` for everything else.
+`object.name` sends `INVOKE`.
 
-`object.to_a` and `object.to_ary` calls `Truffle::Interop.to_array(object)`
+`object.name(*args)` sends `INVOKE`.
+
+`object.new(*args)` sends `NEW`.
+
+`object.to_a` and `object.to_ary` calls `Truffle::Interop.to_array(object)`.
```

Comment:
`Polyglot.to_array` instead?

Reply:
No the new polyglot API is just a wrapper around `Truffle::Interop`, so the implementation methods are often the same as they were before. Since this document is now in `contributor` rather than `user` I'm leaving these implementation details documented here.

Reply:
ok thanks. (sorry for short message, I wanted to make a note to myself to recheck it when I read the whole thing, but I've posted it as a single comment instead by mistake)